server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/WebSocketRoutes.scala

```diff
@@ -0,0 +1,145 @@
+      .action(this.handleWebSockets)
+      .corsHeaders,
+    JMAPRoute.builder
+      .endpoint(new Endpoint(HttpMethod.OPTIONS, JMAP_WS))
+      .action(JMAPRoutes.CORS_CONTROL)
+      .corsHeaders())
+
+  private def handleWebSockets(httpServerRequest: HttpServerRequest, httpServerResponse: HttpServerResponse): Mono[Void] = {
+    SMono(authenticator.authenticate(httpServerRequest))
+      .flatMap((mailboxSession: MailboxSession) => userProvisioner.provisionUser(mailboxSession)
+        .`then`
+        .`then`(SMono(httpServerResponse.sendWebsocket((in, out) => handleWebSocketConnection(mailboxSession)(in, out)))))
+      .onErrorResume(throwable => handleHttpHandshakeError(throwable, httpServerResponse))
+      .subscribeOn(Schedulers.elastic)
+      .asJava()
+      .`then`()
+  }
+
+  private def handleWebSocketConnection(session: MailboxSession)(in: WebsocketInbound, out: WebsocketOutbound): Mono[Void] =
+    SFlux[WebSocketFrame](in.aggregateFrames()
+      .receiveFrames())
+      .map(frame => {
+        val bytes = new Array[Byte](frame.content().readableBytes)
+        frame.content().readBytes(bytes)
+        new String(bytes, StandardCharsets.UTF_8)
+      })
+      .flatMap(handleClientMessages(session))
+      .onErrorResume(e => SMono.just(asError(None)(e)))
+      .map(ResponseSerializer.serialize)
+      .map(_.toString)
+      .flatMap(response => out.sendString(SMono.just(response), StandardCharsets.UTF_8))
+      .onErrorResume(e => {
+        e.printStackTrace()
+        SMono.empty
+      })
+      .`then`()
+      .asJava()
+      .`then`()
+
+  private def handleClientMessages(session: MailboxSession)(message: String): SMono[WebSocketOutboundMessage] =
+    ResponseSerializer.deserializeWebSocketInboundMessage(message)
+      .fold(invalid => {
+        val error = asError(None)(new IllegalArgumentException(invalid.toString()))
+        SMono.just[WebSocketOutboundMessage](error)
+      }, {
+          case request: WebSocketRequest =>
+            jmapApi.process(request.requestObject, session)
+              .map[WebSocketOutboundMessage](WebSocketResponse(request.requestId, _))
+              .onErrorResume(e => SMono.just(asError(request.requestId)(e)))
+              .subscribeOn(Schedulers.elastic)
+        })
+
+  private def handleHttpHandshakeError(throwable: Throwable, response: HttpServerResponse): SMono[Void] = throwable match {
+    case e: UnauthorizedException =>
+      LOGGER.warn("Unauthorized", e)
+      respondDetails(response,
+        ProblemDetails(status = UNAUTHORIZED, detail = e.getMessage),
+        UNAUTHORIZED)
+    case e =>
+      LOGGER.error("Unexpected error upon WebSocket handshake request", e)
+      respondDetails(response,
+        ProblemDetails(status = INTERNAL_SERVER_ERROR, detail = e.getMessage),
+        INTERNAL_SERVER_ERROR)
+  }
+
+  private def asError(requestId: Option[RequestId])(throwable: Throwable): WebSocketError = throwable match {
```

Comment:
That's quite a bunches of duplicates.

I wonder if a `ProblemDetails.of(throwable)` would not allow mutualizing error handling between the HTTP and the websocket stack...